include/kernel.h

```diff
@@ -0,0 +1,534 @@
+	/** Mailboxes attached to this event */
+	kmbox_t *mboxes[MAX_MAILBOXES_PER_EVENT];
+} kevent_t;
+
+
+/** @brief Event generated when a SP interrupt is triggered */
+extern kevent_t KEVENT_IRQ_SP;
+/** @brief Event generated when a SI interrupt is triggered */
+extern kevent_t KEVENT_IRQ_SI;
+/** @brief Event generated when a AI interrupt is triggered */
+extern kevent_t KEVENT_IRQ_AI;
+/** @brief Event generated when a VI interrupt is triggered */
+extern kevent_t KEVENT_IRQ_VI;
+/** @brief Event generated when a PI interrupt is triggered */
+extern kevent_t KEVENT_IRQ_PI;
+/** @brief Event generated when a DP interrupt is triggered */
+extern kevent_t KEVENT_IRQ_DP;
+/** @brief Event generated when the RESET button is pushed (pre-NMI exception) */
+extern kevent_t KEVENT_RESET;
+
+
+/** 
+ * @brief Initialize the multi-threading kernel
+ *
+ * The current execution context becomes the main thread of the
+ * program, with priority set to 0 (you can change priority
+ * to the main thread as well using #thread_set_pri).
+ *
+ * The main thread uses the original stack allocated for the
+ * whole process, so it is technically unbounded (or limited by
+ * heap size).
+ *
+ * @return a pointer to the main thread.
+ *
+ */
+kthread_t* kernel_init(void);
+
+/** 
+ * @brief Shutdown the multi-threading kernel
+ *
+ * This function is mostly useful for testing purposes.
+ * Since the kernel does not keep track of all created threads,
+ * this function should be called only when all created threads
+ * are exited or have been killed.
+ */
+void kernel_close(void);
+
+/**
+ * @brief Create a new thread
+ *
+ * Create a new thread, with a specified stack size and priority.
+ * The thread is immediately made ready after creation, so if it has
+ * a priority higher than or equal to the current thread, it will be
+ * scheduled immediately, before thread_new returns.
+ *
+ * The thread will begin execution from the specified entry point
+ * function. If the function ever returns, the thread is automatically
+ * killed.
+ *
+ * @param[in] name
+ *            Name of the thread (for debugging purposes)
+ * @param[in] stack_size
+ *            Size of the stack in bytes. Minimum suggested size
+ *            is 2048.
+ * @param[in] pri
+ *            Priority of the thread (0-127). Negative values are not allowed.
+ *            Higher number means higher priority.
+ * @param[in] user_entry
+ *            Entry point of the thread. This is the function that will
+ *            be called when the thread begins execution. If this function
+ *            ever returns, the thread is automatically killed.
+ * @param[in] user_data
+ *            Argument that will be passed to the entry point of the thread.
+ *
+ * @return    A pointer to the new thread. It is not necessary to store
+ *            this reference if not required; the thread will clean up
+ *            after itself when it exits.
+ */
+kthread_t* kthread_new(const char *name, int stack_size, int8_t pri,
```

Comment:
I'd vote for removing the name from here. I think we can make this more generic without a name. Consider;

        typedef struct kudata_s
        {
	        uint8_t id;
                const char *name;
                // anything I want as a consumer
        } kudata_t;
        
        kudata_t thread_data_1 = { .id = 0, .name="thread1" };
        kudata_t thread_data_2 = { .id = 0, .name="thread2" };
        
        kthread_new(2048, 2, thread_entry, &thread_data_1);
        kthread_new(2048, 2, thread_entry, &thread_data_2);

If we need the name as a lower-level debugging tool, we can enable it behind a define IMO.

Reply:
This kind of construct requires the user to define a structure to be able to pass custom data to the thread entry point. Let's say I have a audio entry point which is `play_sound_effect` and I have `sound_t` as descriptor for a sound. Right now, I can do something like `kthread_new("audio", 4096, PRI_AUDIO, play_sound_effect, (void*)my_sound);`. With your proposal  I need to define my own `kudata_sound_effect_t` structure which inherits `kudata_t` and contains an additional field of a type `sound_t*` and then pass it `kthread_new`. It looks too much verbose to me.

Reply:
Agreed. Knowing `name` is also integrated with the current implementation, this is just fine. I originally thought it was for user code debugging and was superficial. Maybe we can note the name is for integrated debugging extensions.